### ISSUE TITLE
[CSS+JS] idea for implementing the @document rule for many browsers

### ISSUE THREAD

**Issue by An-Error94** [mdn/sprints#2661]
# Request type

- [x] Correction or update

# Details

I am writing about the change I tried to make in the document: https://developer.mozilla.org/en-US/docs/Web/CSS/@document on January 6, 2020.
My goal was to show how to replace the @document rule in combination with function domain() in user stylesheet (in userContent.css) so that it can be used in any web browser. The @document rule is only supported in firefox, while the attribute selector is widely supported. Unfortunately, my idea also uses JavaScript, so you'll also need a user script (run in firefox using an add-on). I saw no other way to replace the @document rule currently unsupported in other browsers. I'm sick of waiting. The @document rule (in the form @-moz-document) is supported in firefox since version 1.5, i.e. for about 14 years. Over the past 14 years, other browser developers have not implemented their counterpart to this very useful (at least in user style sheets) rule! Earlier, I was still not sure if the selector was properly selected, i.e. whether it works only on selected pages. This method, which I described below, is in my opinion very clever and transparent. However, I didn't know how to describe it on MDN's page (https://developer.mozilla.org/en-US/docs/Web/CSS/@document). I don't write fluent English either, so I'd like someone to describe my idea on this page. I think this site is suitable for this purpose - at least in user scripts - to use solutions that will work on many web browsers. My idea is as follows:
JavaScript reads the full domain name and assigns this string as a custom data attribute to an HTML element that is higher than the topmost HTML elements that can be referenced in CSS selectors. In my opinion, this element should be 'body'. In other words, the 'body' element will get an attribute whose value will be the full domain name of the server where this HTML document is. So the "body" element on this page (https://developer.mozilla.org/en-US/docs/Web/CSS/@document) would look like this: 'body[data-at-document="developer.mozilla.org"]'
The "data-at-document" attribute is a custom data attribute named "at-document" because the "data-" part is mandatory for this type of attribute.
The JavaScript code implementing the above-mentioned action looks like this:
if(document.location.hostname != ""){
document.querySelector('body').setAttribute('data-at-document', document.location.hostname);
}
The "if" statement is used to exclude situations where the string obtained from "document.location.hostname" is empty. This happens e.g. when a saved page from the internet is opened on the user's computer. If the "data-at-document" attribute was then assigned an empty string value, this would prevent the use of the selectors that were used when the page was on the server.
After these actions, the @document rule in conjunction with the domain() function is replaced with a CSS selector, where instead of "domainString" enter a string representing the domain name:
'body[data-at-document$="domainString"] '
At the end of this selector there is a space that allows you to select any descendant of the body element if 'body' has the "data-at-document" attribute and the value of this attribute is suffixed (followed) by "domainString", e.g.
body[data-at-document$="mozilla.org"] a
selects each anchor on the pages of the domain "mozilla.org" and all its subdomains
However, there is a problem because 'body[data-at-document$="mozilla.org"] ' will also work on "badmozilla.org"
You can then use the exact comparison (without the "$" sign), but then no subdomain will be included:
body[data-at-document="mozilla.org"] div
selects each element of 'div' on the pages of the domain "mozilla.org"

This idea to bypass the failure of other web browsers to implement the @document rule can be improved. You can use "document.URL" and "[attr ^ = value]" to replace the @document rule and the "url ()" and "url-prefix ()" functions. What I wrote here is just a sketch. I can try to develop it, but I don't want to waste time. What I came up with, I wrote here and it is absolutely enough for me.

I would like to overcome the inability to wait for the @document rule to be implemented in other browsers.

**Comment by chrisdavidmills** [comment]
Hi there @An-Error94! 

Thank you for providing this explanation — this makes a lot more sense now.

So, I think it is valuable to share this somewhere, but I'm not sure directly on the MDN page is the right place. It is basically a polyfill, and we made a decision a while ago to discourage polyfills being shared on MDN, because that gives us another set of tools to have to maintain, and we just don't have the time.

But I would be in favor of helping you out in other ways. Perhaps you could write it up and put it on a Github repo somewhere? Then we could link to it from MDN.

I'd also be happy to help edit the language.

**Comment by An-Error94** [comment]
Hi @chrisdavidmills!

Thank you for your response. Thank you very also much for the offer of help on your part.

I really like your suggestion. I will write the code in my GitHub repository. The link to it is as follows: https://github.com/An-Error94/Handy-Scripts/tree/master/%40document%20polyfill

I'm afraid I will only be able to create a solution for "url ()", "url-prefix ()" and "domain ()".

**Comment by chrisdavidmills** [comment]
Cool, let me know when it is ready and I will copy edit it.

**Comment by An-Error94** [comment]
Thank you so much for wanting to help me.

I have already finished my polyfill. I need an external opinion about this polyfill. I did what I could. I think it can't be done better. I don't know if I chose a good substitute character for the hyphen (I chose "^" - it goes to the attribute) in the domain () function. I tried to write legibly and correctly in English, but in this matter I need help if something would be incomprehensible. I abandoned the idea with the attribute matching the end of the attribute value (attribute $ = "value") due to the problems mentioned in the previous message.

I am not sure if the current result is suitable for publication.
